### PR TITLE
Adds SWE notes using mkdocs

### PR DESCRIPTION
**Pull request by bishtgautam**
Adds equations for SWE+Roe solver based on Kim et al. (2012)

Kim, J., Warnock, A., Ivanov, V. Y., & Katopodes, N. D. (2012).
Coupled modeling of hydrologic and hydrodynamic processes including
overland and channel flow. Advances in water resources, 37, 104-126.
https://www.sciencedirect.com/science/article/pii/S0309170811002211?via%3Dihub

The markdown notes are available on https://rdycore.github.io/toy-problems/

### CODE REVIEW

**Comment by bishtgautam** [comment]
@donghuix Can you please review the equations for accuracy?
 - Raw markdown file is https://github.com/RDycore/toy-problems/pull/2/files#diff-791da51d6bc3d10545749bfcf8b1de8e1d1dd4a9787bd5441cfab08826d760bb
 - Rendered markdown file is https://rdycore.github.io/toy-problems/shallow_water_equation/
 
 @jeff-cohere Can you please review the PR for the use of mkdocs?

File: docs/shallow_water_equation.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+$\mathbf{F}$ is the flux vector and
+$\mathbf{n}$ is the outward pointing unit vector to the boundary $\partial\Omega$. 
+The flux normal flux across the face, $\mathbf{F.n}$, is given by
+
+\begin{align}
+  \mathbf{F.n}
+  =
+  \begin{bmatrix}
+  hu_\perp  \\[.5em]
+  huu_\perp + \frac{1}{2}gh^2 \cos\phi \\[.5em]
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: There is another term to be added : /dfrac{1}{24}g(\Delta h)^{2} \cos \phi

File: docs/shallow_water_equation.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  hvu_\perp + \frac{1}{2}gh^2 \sin\phi
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: + /dfrac{1}{24}g(\Delta h)^{2} \sin \phi

File: docs/shallow_water_equation.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+\end{equation}
+
+where
+
+\begin{align}
+  \mathbf{R}
+  =
+  \begin{bmatrix}
+  1 & 0 & 1  \\[.5em]
+  \hat{u} - \hat{c}\cos\phi & -\sin\phi & \hat{u} + \hat{c}\cos\phi \\[.5em]
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: I think it should be \hat{a} instead of \hat{c}.

File: docs/shallow_water_equation.md [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+  \end{bmatrix}
+\end{align}
+
+The asterisks denote that the eigenvalues 
+$\hat{\lambda}_1 (= \hat{u}_\perp - \hat{a} )$ and
+$\hat{\lambda}_3 (= \hat{u}_\perp + \hat{a})$ 
+are adjusted since Roe's method does not provide correct flux for critical flow.
+
+\begin{eqnarray}
+  |\hat{\lambda}|_1 &=& \frac{\hat{\lambda}^2_1}{\Delta \lambda} + \frac{\Delta \lambda}{4} \mbox{\hspace{0.5cm}if $-\Delta \lambda/2 < \hat{\lambda}_1 < \Delta \lambda/2$} \\
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: \hspace{0.5cm} is not interpreted.

File: docs/shallow_water_equation.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+\begin{eqnarray}
+\int_{d\Omega} \mathbf{S} d\Omega &=& \int_{\Omega} \left( -gh\frac{\partial z}{\partial x} - C_D u \sqrt{u^2 + v^2} \right) d\Omega \nonumber \\
+&=& \int_{d\Omega}  -gh\frac{\partial z}{\partial x} d\Omega - \int_{\Omega} C_D u \sqrt{u^2 + v^2}  d\Omega 
+\end{eqnarray}
+
+#### Bed slope elevation term
+
+\begin{eqnarray}
+\int_{d\Omega}  -gh\frac{\partial z}{\partial x} d\Omega &\approx& -gh\overline{\frac{\partial z}{\partial x}} \Omega \nonumber \\[0.6em]
+&=& \frac{(y_2 - y_0)(z_1 - z_0) - (y_1 - y_0)(z_2 - z_0)}{(y_2 - y_0)(x_1 - x_0) - (y_1 - y_0)(x_2 - x_0)} \Omega
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: We need to specify this is for triangle grid cell.

File: docs/shallow_water_equation.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+#### Roughness term
+
+\begin{eqnarray}
+\int_{\Omega} C_D u \sqrt{u^2 + v^2}  d\Omega  &\approx& C_D u \sqrt{u^2 + v^2} \Omega
+\end{eqnarray}
+
+## PETSc TS implementation
+
+The PETSc TS notation\footnote{https://petsc.org/release/docs/manual/ts/} for solving time-dependent is given as:
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: notation\footnote{https://petsc.org/release/docs/manual/ts/}  is not interpreted.

File: docs/shallow_water_equation.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+\begin{equation}
+\label{eqn:petsc_ts}
+\mathtt{ F(t,u,{u}_t)=G(t,u)}
+\end{equation}
+
+By comparing \ref{eqn:swe} and \ref{eqn:petsc_ts},
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: \ref{eqn:swe} and \ref{eqn:petsc_ts} are not interpreted in latex.

**Review by donghuix — COMMENTED** [review]
Please find some typos and issues in the code.

File: docs/shallow_water_equation.md (hunk 6 above)
Comment (jeff-cohere) [reply docs/shallow_water_equation.md]: The "idiomatic" way to reference links on a webpage (as opposed to a journal article) is just to use a hyperlink, like

```
The [PETSc TS notation](https://petsc.org/release/docs/manual/ts/) for solving time-dependent equations is given as:
...
```

File: docs/shallow_water_equation.md (hunk 7 above)
Comment (jeff-cohere) [reply docs/shallow_water_equation.md]: MkDocs (and probably Sphinx too) use [MathJax](https://docs.mathjax.org/en/latest/) for rendering mathematical formulae. According to [this](https://docs.mathjax.org/en/latest/input/tex/extensions/textmacros.html?highlight=eqref#equation-numbers), `\ref` and `\eqref` are supported in the latest spec, but I'm not sure we're using that.

File: docs/shallow_water_equation.md (hunk 4 above)
Comment (jeff-cohere) [reply docs/shallow_water_equation.md]: See [this](https://docs.mathjax.org/en/latest/input/tex/macros/index.html) for supported LaTeX commands.

File: docs/shallow_water_equation.md (hunk 2 above)
Comment (bishtgautam) [reply docs/shallow_water_equation.md]: This third term is in Kim et al. (2012) but isn't in the series of papers by Sanders. I can't figure out which paper derived this third term. Do you know?

File: docs/shallow_water_equation.md (hunk 3 above)
Comment (bishtgautam) [reply docs/shallow_water_equation.md]: fixed

File: docs/shallow_water_equation.md (hunk 5 above)
Comment (bishtgautam) [reply docs/shallow_water_equation.md]: done

File: docs/shallow_water_equation.md (hunk 6 above)
Comment (bishtgautam) [reply docs/shallow_water_equation.md]: done

File: docs/shallow_water_equation.md (hunk 7 above)
Comment (bishtgautam) [reply docs/shallow_water_equation.md]: done

File: docs/shallow_water_equation.md (hunk 2 above)
Comment (bishtgautam) [reply docs/shallow_water_equation.md]: Okay, I figured out the origin of the third term. It is eq 13 in Bradford and Sanders (2002) paper (https://ascelibrary.org/doi/10.1061/%28ASCE%290733-9429%282002%29128%3A3%28289%29)

File: docs/shallow_water_equation.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+\end{equation}
+
+where
+
+\begin{align}
+  \mathbf{R}
+  =
+  \begin{bmatrix}
+  1 & 0 & 1  \\[.5em]
+  \hat{u} - \hat{a}\cos\phi & -\sin\phi & \hat{u} + \hat{c}\cos\phi  \\[.5em]
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: It should be \hat{u} + \hat{a}\cos\phi in the last term. I just realized \hat{c} is used in Bradford et al. 2002 and \hat{a} is used in Kim et al 2012. Let's be consistent here.

File: docs/shallow_water_equation.md [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+  \hat{v} - \hat{a}\sin\phi & \cos\phi & \hat{v} + \hat{c}\sin\phi
```
Comment (donghuix) [inline docs/shallow_water_equation.md]: \hat{v} + \hat{a}\sin\phi

**Review by donghuix — COMMENTED** [review]
There is still a typo.

**Comment by tanzeli1982** [comment]
@donghuix Will the added equations be shown at the latex format after PR is merged? Currently, they are still shown at a text format.

**Comment by jedbrown** [comment]
@tanzeli1982 They'll show up on this page. https://rdycore.github.io/toy-problems/shallow_water_equation/

@bishtgautam @jeff-cohere We could make a test deployment so there is a place to read rendered docs within the pull request. Or maybe that much automation isn't needed at this time.

**Comment by bishtgautam** [comment]
Adding automated rendering of docs for each PR is a good idea. I have now added [`.github/workflows/auto_publish_mkdocs.yml`](https://github.com/RDycore/toy-problems/pull/2/files#diff-9d231963f3fd1ac81acdd997ef05f711ec9932b7dd1dbbb20b842c0aa2dae984R1
) but Github Actions isn't picking it up.

@jedbrown @jeff-cohere I have given you admin permission for this repo. Do you have an idea why Github Actions is running the `auto_publish_mkdocs.yml`?

File: docs/shallow_water_equation.md (hunk 8 above)
Comment (bishtgautam) [reply docs/shallow_water_equation.md]: Good catch. I have fixed it now.

**Comment by bishtgautam** [comment]
Okay, I just realized that the documentation will be built once the PR is merged and not when the PR is issued based on the following settings:

```
on:
  push:
    branches:
      - main
  
```

I think this setup is okay. What do you think?

**Comment by jedbrown** [comment]
I think that's okay. GitLab-CI supports separate deployed environments within a merge request (they're intended for review), but I don't think GitHub Actions has an equivalent concept yet (it might be possible to add it somewhere within the existing site, or we could make a GitHub repository for reviews). ReadTheDocs does something like that and reports back to the GitHub issue.

**Review by donghuix — APPROVED** [review]
It looks good to me.

**Comment by bishtgautam** [comment]
Ok second thoughts, I updated the GitHub Actions to generate the documentation on opening the PR as it is easier to look at the rendered equations on https://rdycore.github.io/toy-problems than reading the `.md` when reviewing the PR.

@jeff-cohere Do you have any additional comments for this PR?

**Comment by jedbrown** [comment]
Does this mean that opening a PR will replace the docs that have been generated on `main` or any other PR? That seems not desirable in general (especially once the site gets visitors beyond the development team). Maybe we can put it in `tdycore.github.io/toy-problems/pr123/` if there isn't a better solution purely within GitHub actions.

**Comment by jeff-cohere** [comment]
I agree with Jed that we probably don't wanr to replace the docs for every PR. If we could generate them in a unique subdirectory related to the PR, that would solve the issue.

The "low tech" way to address this problem would be for reviewers to run mkdocs serve on the branch on their machine to see the files. That's not hard, but the PR-specific link would be easier.

**Comment by jeff-cohere** [comment]
(Other than this, the PR looks good to me!)

**Comment by bishtgautam** [comment]
Agreeing with @jedbrown and @jeff-cohere, I'm reverting the last commit.
